File: examples/typescript/notion/notion.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { resolve } from 'node:path'
import { fileURLToPath } from 'node:url'
import dotenv from 'dotenv'
import { MountMode, NotionResource, Workspace, type NotionConfig } from '@struktoai/mirage-node'

const __HERE = fileURLToPath(new URL('.', import.meta.url))
dotenv.config({ path: resolve(__HERE, '../../../.env.development') })

function buildConfig(): NotionConfig {
  const apiKey = process.env.NOTION_API_KEY
  if (apiKey === undefined || apiKey === '') {
    throw new Error('NOTION_API_KEY env var is required')
  }
  return { apiKey }
}

async function run(ws: Workspace, cmd: string): Promise<string> {
  console.log(`$ ${cmd}`)
  const r = await ws.execute(cmd)
  if (r.exitCode !== 0 && r.stderrText !== '') {
    console.log(`  STDERR: ${r.stderrText.slice(0, 200)}`)
  }
  const out = r.stdoutText.replace(/\s+$/, '')
  if (out !== '') {
    for (const line of out.split('\n').slice(0, 10)) console.log(`  ${line.slice(0, 200)}`)
  }
  return out
}

async function main(): Promise<void> {
  const ws = new Workspace(
    { '/notion': new NotionResource(buildConfig()) },
    { mode: MountMode.READ },
  )
  try {
    console.log('=== ls /notion/pages/ ===')
    const p0 = (await run(ws, 'ls /notion/pages/ | head -n 1')).trim()
    if (p0 === '') {
      console.log('no pages')
      return
    }
    const pagePath = `/notion/pages/${p0}`

    console.log(`\n=== cat ${pagePath}/page.json ===`)
    await run(ws, `cat "${pagePath}/page.json"`)

    console.log(`\n=== jq .title ${pagePath}/page.json ===`)
    await run(ws, `jq ".title" "${pagePath}/page.json"`)

    console.log(`\n=== jq .url ${pagePath}/page.json ===`)
    await run(ws, `jq ".url" "${pagePath}/page.json"`)

    console.log(`\n=== jq .markdown ${pagePath}/page.json ===`)
    await run(ws, `jq ".markdown" "${pagePath}/page.json"`)

    console.log(`\n=== stat ${pagePath}/page.json ===`)
    await run(ws, `stat "${pagePath}/page.json"`)

    console.log(`\n=== head -n 5 ${pagePath}/page.json ===`)
    await run(ws, `head -n 5 "${pagePath}/page.json"`)

    console.log('\n=== tree -L 1 /notion/ ===')
    await run(ws, 'tree -L 1 /notion/')

    console.log(`\n=== tree -L 1 ${pagePath}/ ===`)
    await run(ws, `tree -L 1 "${pagePath}/"`)

    console.log(`\n=== find ${pagePath}/ -name '*.json' ===`)
    await run(ws, `find "${pagePath}/" -name "*.json"`)

    console.log(`\n=== basename ${pagePath}/page.json ===`)
    await run(ws, `basename "${pagePath}/page.json"`)

    console.log(`\n=== dirname ${pagePath}/page.json ===`)
    await run(ws, `dirname "${pagePath}/page.json"`)

    console.log('\n=== notion-search --query EVO ===')
    await run(ws, 'notion-search --query EVO')

    console.log(`\n=== grep Graph ${pagePath}/*.json (page glob) ===`)
    await run(ws, `grep -c Graph "${pagePath}/"*.json`)

    console.log('\n=== rg Graph /notion/pages/ ===')
    await run(ws, 'rg -c Graph /notion/pages/')

    console.log(`\n=== ls ${pagePath}/ (children) ===`)
    const children = await run(ws, `ls "${pagePath}/"`)
    const childDirs = children
      .trim()
      .split('\n')
      .filter((line) => line !== '' && !line.endsWith('.json'))
    if (childDirs.length > 0) {
      const child = childDirs[0]!
      console.log(`\n=== jq .title ${pagePath}/${child}/page.json (child page) ===`)
      await run(ws, `jq ".title" "${pagePath}/${child}/page.json"`)
    }

    console.log(`\n=== echo ${pagePath}/*.json (glob) ===`)
    await run(ws, `echo "${pagePath}/"*.json`)

    const records = ws.records
    const total = records.reduce((acc, r) => acc + (r.bytes ?? 0), 0)
    console.log(`\nStats: ${String(records.length)} ops, ${String(total)} bytes transferred`)
  } finally {
    await ws.close()
  }
}

main().catch((err: unknown) => {
  console.error(err)
  process.exit(1)
})

File: examples/python/notion/notion.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
import os

from dotenv import load_dotenv

from mirage import MountMode, Workspace
from mirage.resource.notion import NotionConfig, NotionResource

load_dotenv(".env.development")

config = NotionConfig(api_key=os.environ["NOTION_API_KEY"])
resource = NotionResource(config=config)


async def main() -> None:
    ws = Workspace({"/notion": resource}, mode=MountMode.READ)

    print("=== ls /notion/pages/ ===")
    result = await ws.execute("ls /notion/pages/")
    print(await result.stdout_str())

    first_page = (await result.stdout_str()).strip().splitlines()[0] if (
        await result.stdout_str()).strip() else ""
    if not first_page:
        print("No pages available")
        return

    page_path = f"/notion/pages/{first_page}"

    print(f"=== cat {page_path}/page.json ===")
    result = await ws.execute(f"cat {page_path}/page.json")
    print((await result.stdout_str())[:2000])

    print(f"\n=== jq .title {page_path}/page.json ===")
    result = await ws.execute(f'jq ".title" {page_path}/page.json')
    print(await result.stdout_str())

    print(f"=== jq .url {page_path}/page.json ===")
    result = await ws.execute(f'jq ".url" {page_path}/page.json')
    print(await result.stdout_str())

    print(f"=== jq .markdown {page_path}/page.json ===")
    result = await ws.execute(f'jq ".markdown" {page_path}/page.json')
    print((await result.stdout_str())[:500])

    print(f"\n=== stat {page_path}/page.json ===")
    result = await ws.execute(f"stat {page_path}/page.json")
    print(await result.stdout_str())

    print(f"=== head -n 5 {page_path}/page.json ===")
    result = await ws.execute(f"head -n 5 {page_path}/page.json")
    print(await result.stdout_str())

    print("=== tree -L 1 /notion/ ===")
    result = await ws.execute("tree -L 1 /notion/")
    print(await result.stdout_str())

    print(f"=== tree -L 1 {page_path}/ ===")
    result = await ws.execute(f"tree -L 1 {page_path}/")
    print(await result.stdout_str())

    print(f"=== find {page_path}/ -name '*.json' ===")
    result = await ws.execute(f'find {page_path}/ -name "*.json"')
    print(await result.stdout_str())

    print(f"=== basename {page_path}/page.json ===")
    result = await ws.execute(f"basename {page_path}/page.json")
    print(await result.stdout_str())

    print(f"=== dirname {page_path}/page.json ===")
    result = await ws.execute(f"dirname {page_path}/page.json")
    print(await result.stdout_str())

    print("=== notion-search --query EVO ===")
    result = await ws.execute("notion-search --query EVO")
    print((await result.stdout_str())[:1000])

    print(f"\n=== grep Graph {page_path}/*.json (page glob) ===")
    result = await ws.execute(f'grep -c Graph "{page_path}/"*.json')
    out = (await result.stdout_str()).strip()
    lines = out.splitlines() if out else []
    print(f"  exit={result.exit_code} matches: {len(lines)}")
    for line in lines[:3]:
        print(f"  {line[:150]}")

    print("\n=== rg Graph /notion/pages/ ===")
    result = await ws.execute('rg -c Graph /notion/pages/')
    out = (await result.stdout_str()).strip()
    lines = out.splitlines() if out else []
    print(f"  exit={result.exit_code} matches: {len(lines)}")
    for line in lines[:3]:
        print(f"  {line[:150]}")

    print(f"\n=== ls {page_path}/ (children) ===")
    result = await ws.execute(f"ls {page_path}/")
    children = await result.stdout_str()
    print(children)

    child_dirs = [
        line for line in children.strip().splitlines()
        if not line.endswith(".json")
    ]
    if child_dirs:
        child = child_dirs[0]
        print(f"=== cat {page_path}/{child}/page.json (child page) ===")
        result = await ws.execute(f"cat {page_path}/{child}/page.json")
        print((await result.stdout_str())[:1000])

    # ── glob expansion (exercises resolve_glob → readdir) ──
    print(f"\n=== echo {page_path}/*.json (glob) ===")
    r = await ws.execute(f"echo {page_path}/*.json")
    out = (await r.stdout_str()).strip()
    print(f"  {out[:200]}")


if __name__ == "__main__":
    asyncio.run(main())
